Open space port (Multi-Z view)

CrimsonShrike opened the pull request:
This is a port from vorestations port from Eris of openspace multi z view.

I believe it's pretty much finalized. I will hold on for features such as look up and climbing from Eris when it makes enough sense on map.


<!-- 
Do not forget to add a changelog when you have made admin/player facing changes that can alter gameplay.
Examples which require a changelog entry include:
* Adding/removing objects that players may interact with, or the way they function.
* Adding/removing/altering admin tools.
* Changing the map.

Examples were changelog entries are optional/not typically required:
* Cosmetic changes such as descriptions, sound effects, etc.
* Optimizations and other changes to underlying systems which do not affect gameplay.
* Minor bug fixes.

You find a README and example file in .\html\changelogs\ for further instructions.
-->

mustafakalash commented on code/game/turfs/space/space.dm:
```diff
@@ -212,3 +212,6 @@
 	icon_state = "bluespace"
 	keep_sprite = 1
+
+/turf/space/is_space()
```
we have https://github.com/CrimsonShrike/Baystation12/blob/dev/code/_macros.dm#L54

mustafakalash commented on code/modules/multiz/zshadow.dm:
```diff
@@ -0,0 +1,116 @@
+		shadow.set_dir(new_dir)
+
+
+// We should show the typing indicator so people above us can tell we're about to talk.
```
Don't leave commented out code in the final version

CrimsonShrike commented on test/run-test.sh:
```diff
@@ -240,7 +240,7 @@ function run_byond_tests {
     fi
     run_test_ci "check globals build" "python tools/GenerateGlobalVarAccess/gen_globals.py baystation12.dme code/_helpers/global_access.dm"
-    run_test "check globals unchanged" "md5sum -c - <<< '219da80dfc00a6b58952b6afd8b104ab *code/_helpers/global_access.dm'"
+    run_test "check globals unchanged" "md5sum -c - <<< '92eeed2ee3095302c9c375609f533f13 *code/_helpers/global_access.dm'"
```
May have misunderstood what Psy meant and touched something I shouldn't.

mustafakalash commented:
You need to use the new global variable system, rather than editing the unit test.

CrimsonShrike commented:
Latest commit fixes some things but the method to it isn't particularly good, must find a better way to keep overlays from inferior levels in place.

CrimsonShrike commented on code/controllers/subsystems/open_space_controller.dm:
```diff
@@ -3,7 +3,6 @@
 //
 
 GLOBAL_VAR_INIT(open_space_initialised, FALSE)
-GLOBAL_DATUM_INIT(OS_controller, /datum/controller/subsystem/open_space, null)
```
Since I am creating a global with the subsystem this seemed redundant, point out if it's bad practice to make use of global instead of GLOB in this case though.

CrimsonShrike commented:
Dangit, I broke something toying with GK. DNM

Techhead0 commented on code/controllers/subsystems/open_space_controller.dm:
```diff
@@ -0,0 +1,112 @@
+	name = "openspace"
+	init_order = INIT_OPEN_SPACE
+	var/list/turfs_to_process = list()		// List of turfs queued for update.
+	var/list/turfs_to_process_old = null	// List of turfs currently being updated.
```
If `turfs_to_process_old` is only used by `fire()`, does it need to be stored as a var?

PsiOmegaDelta commented on code/modules/multiz/zshadow.dm:
```diff
@@ -0,0 +1,99 @@
+	return owner.hear_say(message, verb, language, alt_name, italics, speaker, speech_sound, sound_vol)
+
+/mob/zshadow/proc/sync_icon(var/mob/M)
+	name = M.name
```
Store vars you wish to keep unchanged, then do `appearance = M`, and then reapply the stored vars.

PsiOmegaDelta commented on code/controllers/subsystems/open_space_controller.dm:
```diff
@@ -0,0 +1,111 @@
+		if(isopenspace(T))
+			// log_debug("[T] ([T.x],[T.y],[T.z]) queued for update for [src]New()")
+			OS_controller.add_turf(T, 1)
+*/
```
Don't keep dead code.

PsiOmegaDelta commented on code/controllers/subsystems/open_space_controller.dm:
```diff
@@ -0,0 +1,111 @@
+		var/turf/T = GetAbove(src)
+		if(isopenspace(T))
+			// log_debug("[T] ([T.x],[T.y],[T.z]) queued for update for [src].Exited([AM])")
+			SSopen_space.add_turf(T, 1)
```
Did you ever plan to unregister turfs, or is this subsystem supposed to eventually be listening to every turf in the world?

PsiOmegaDelta commented on code/modules/multiz/turf.dm:
```diff
@@ -54,31 +55,79 @@
+		var/depth = 1
+		for(var/T = GetBelow(src); isopenspace(T); T = GetBelow(T))
+			depth += 1
+			to_chat(user, "It is about [depth] levels deep.")
```
`"It is about [depth] level\s deep."`

PsiOmegaDelta commented on code/modules/multiz/turf.dm:
```diff
@@ -54,31 +55,79 @@
+
+		var/overlays_post = overlays.len
+		if(overlays_post != (overlays_pre + o_img.len)) //Here we go!
+			log_world("Corrupted openspace turf at [x],[y],[z] being replaced. Pre: [overlays_pre], Post: [overlays_post]")
```
Lingering logs.

PsiOmegaDelta commented on code/controllers/subsystems/open_space_controller.dm:
```diff
@@ -0,0 +1,111 @@
+		A.fall()
+	T.update_icon()
+
+/datum/controller/subsystem/open_space/proc/add_turf(var/turf/T, var/recursive = 0)
```
So you're never removing turfs, and turfs can be added multiple times. I'd say this is a major performance concern.

CrimsonShrike replied:
@PsiOmegaDelta  should be handled now, adding has a check and fire() deleted the list of turfs.

BlueNexus commented:
DNM at the author's request.

CrimsonShrike committed:
Initial port commit

Attempts to port Eris open space logic through Vorestation port. Features removed pending reintroduction for initial test.
Remove gotos

Removed gotos, although I think this could be neater an cleaner.
Return null

Fixes runtime errors
Cardemon Tweak -- Stats in Card Name

To make the game actually EASILY playable, not just theoretically.
/obj/fire now returns qdel hint.

Stabilizes Travis.

Medical and reagent splashing tweaks.

Merge pull request #17672 from chaoko99/beauraucracy

Fixes title1
Removes Exodus.

Properly resolves conflict with loadout beterts.

Fixes SDQL2 select and adds more feedback.

Removes comma
Merge pull request #18177 from PsiOmegaDelta/170810-BurnBabyBurn

/obj/fire now returns qdel hint.
Merge pull request #18176 from nearlyNonexistent/cardemon-playable-edition

Cardemon Minor Tweak
Merge pull request #18164 from FearTheBlackout/dev

Adds "Culinary Specialist" as an alternate title for Cook
Merge pull request #18162 from gelza1/patch-1

Updated Alert text
Merge pull request #18168 from Crushtoe/dev

Adds First Deck Observation Lounge
Merge pull request #18182 from mustafakalash/query-whatislove

Fixes SDQL2 select and adds more feedback.
Adds several changes to loadout

- Changes frontier clothes to its own item, as coveralls aren't really a formal outfit.
- Adds Scientist and Senior Researcher to semi-formal roles.
- Adds an "armed role" for jobs that are allowed to carry weaponry as per SOP, or at the very least are allowed to attain them in some capacity (I.E Merchant).
Merge pull request #18170 from PsiOmegaDelta/170809-OptimizesNonRandomness

Re-organizes unit test type overrides.
Adds changelog


Allows cyborgs to use IV drips
- All cyborgs can now eject the blood pack from drip, if adjacent to it.
- Chemistry gripper can now hold blood packs. Technically allows crisis cyborgs to perform transfusions via the IV drip without needing human to swap the blood bags.

Changed actual race information to "Sapient Race"

Added changelog

Removed redundant variables

Merge pull request #18101 from SamSamsonthe3rd/just_kick_it

Adds some cutesy things to vending.
Merge pull request #18090 from Casper3667/organs

allows mechanical livers and kidneys in char setup
Improves changelog


Adds quote to changelog

Kill me now
Human and his brain

Merge pull request #18161 from CrimsonShrike/MultiZLightPort

Introduction to multi-z light
Merge pull request #18179 from mustafakalash/removeexodus

Removes Exodus
Merge pull request #18188 from Superbee29/voicerace

Disabled Tcomms metagaming
Merge pull request #18186 from atlantiscze/2017-08-11-MediborgTweak

Allows cyborgs to use IV drips
Merge pull request #18183 from FearTheBlackout/loadout

Adds several changes to loadout
Merge pull request #18178 from FTangSteve/medStuff

Medical and reagent splashing tweaks.
Merge pull request #18141 from PsiOmegaDelta/170807-KeepYourArtsyAmbitionsAway

Adds a test to check icon state limits.
Merge pull request #18136 from Kearel/deityFormStuff

More deity tweaks
delet this

Merge pull request #18197 from SamSamsonthe3rd/its_time_to_stop

This was a mistake.
Merge branch 'dev' into OpenSpacePort


Multi z WIP

Bringing this out in the open in case anyone wants a go at current issues and to keep code up to standards

-Current issues include lower turf paint being rendered above top turf pipes and things such as vents being rendered on top of objects and mobs
Merge pull request #18223 from AtomicWorm/equal-opportunity-robuster

Changes to security lockers and access.
Makes Bluespace Mirroring dir-consistent

adds hoods to hoodies

Merge pull request #18142 from atlantiscze/2017-08-06-KILL_IT_WITH_FIRE

Removes Canister tgui
Adds extra null check, makes adminpm string code (maybe) more readable.

AltClick Consistency Cleanup

Changelog.

Fixes deity mode bugs.

Hotfix Stack Toxloss

Merge pull request #18229 from Hubblenaut/hotfix

Hotfix Stack Toxloss
Restores on-join announcements for all jobs, with a twist
- Majority of jobs are now announced on respective departmental channel, instead of public channel.
- During red alert, all jobs are announced on main channel regardless of the above. This lets people know when possibly required emergency personnel arrives when situation is seriously screwed up already.
- Vote: https://baystation12.net/forums/threads/vote-restore-on-join-announcements-for-non-command-crew.5091/
- Vote results: 16% for full revert (all jobs announced on main frequency), 60% this (departmental frequencies for most jobs), 24% no change(only few select jobs announced)
- Bonus: Partially/Mostly fixes #18094 - resolves qdel loops caused by overmap edges.

Changelog entry

Merge pull request #18192 from mustafakalash/whowatchesthewatchers

Ticket fixes/changes
Adds a cyborg recharging station to Auxiliary Head


Allows AI to track most faceless mobs via radio again
- Functionally partially reverts #17560
- This has proven to be a MASSIVE annoyance for AI, when everyone is shouting orders to open doors at you, yet you can't track them by clicking the name in chat. You can, however, still track them via the track with camera verb, they are just listed as Unknown (as YOURNAMEHERE). Opening up the verb and picking a target costs you a lot of valuable time, which is annoying when everyone actually expects you to play as an actual AI, multitasking ten things at once.
- This resolves most of the situations where you want to track someone legit (like a security officer or engineer in voidsuit). If you can't track them by face, you will try to track them by worn ID.

Changelog entry

Adds changelog


Merge pull request #18236 from atlantiscze/2017-08-14-LetMeTrackYou

Allows AI to track most faceless mobs via radio again
Merge pull request #18226 from Hubblenaut/bluespace

Makes Bluespace Mirroring dir-consistent
Makes cigs less heart-racing

Now instead of any amount of nicotine increasing pulse, there's a chance based on amount present in bloodstream, 20% at full cig.
Makes overdose increase pulse a lot too.
Lowers overdose from 15 cigs worth to 5.
Adds a define for slimegirls race name


Changes when liver's death spiral kicks in

Instead of bruised level, it now sets in when it's broken.
Self-perpetuating damage is now a bit lower.
Spoons and forks fix
Merge pull request #18240 from Terror4000rus/dev

Spoons and forks in-hand fix
Merge pull request #18235 from FearTheBlackout/stinkytoilet

Adds a cyborg recharging station to Auxiliary Head
Merge pull request #18202 from PsiOmegaDelta/170812-StrictBags

Closets now only store contents during map load.
Merge pull request #18227 from Hubblenaut/fixes

AltClick Consistency Cleanup
Merge pull request #18239 from comma/irongut

Changes when liver's death spiral kicks in
Merge pull request #18233 from atlantiscze/2017-08-13-WelcomeMrSmith

Restores on-join announcements for all jobs, with a twist
Fixes Second Deck Escape Pods not launching

Merge pull request #18232 from Kearel/deityShmeity

Fixes (more) deity mode bugs.
Updates the Lobby and Explosion art.

Special thanks to 50_n00b for creating the art:
https://baystation12.net/forums/threads/the-little-things-4-with-a-vengeance.9/page-50#post-70649

Fixes MMI brain check


Fixes bluegoasts

Explosions make less of a mess with wires

Now only breaks wires if their turf is exposed.
Merge pull request #18253 from Hubblenaut/fixes

Fixes Bluegoasts
Merge pull request #18234 from Hubblenaut/pods

Fixes Second Deck Escape Pods not launching
Merge pull request #18251 from comma/a

Fixes MMI brain check
Merge pull request #18238 from comma/dokidokinicotine

Makes cigs less heart-racing
Fixes #8962 - messenger software gives PDA now

HOTFIX Follow Verb

Fixes Material Duplication Glitch

Microwave sprite pattern fix.

Merge pull request #18265 from Earthcrusher/MicrowaveFix

Microwave sprite pattern fix.
Various Mapping Tweaks

Cleaned the Map

Merge pull request #18231 from Hubblenaut/hotfix2

HOTFIX Follow Verb
Merge pull request #18266 from Crushtoe/i-am-a-failure

Minor Mapping Tweaks and Teleporter Server Crashing Fix
Nerfs lung damage

At some point it was changed so ANY sort of damage makes lung give you 25% of max health in oxygen loss. This was confusing as  lung wasn't considered 'ruptured' by all the checks.
I bumped it back to ruptured lung, and instead made it give you a small amount of oxyloss just to indicate that something's off.
Sets limb blending to additive


Merge pull request #18274 from comma/coughcough

Nerfs lung damage
Merge pull request #18254 from comma/andstaydown

Explosions make less of a mess with wires
Industrial Veil loadout tweak

Just allows Taj to have industrial veils as Engineers, since it was missing before.
Various Map Fixes

Fixes Travis Issues

Merge pull request #18230 from tlc2013/dev

Updates the Lobby and Explosion art.
Fixes #10041 - pAI default language galcom

Fixes #10041 - Galcom default for all synths

Moves proc down to absolute pathing

Fixes indentation error

Adjusts when explosions affect sub-floor objects.

Explosions now only affect atoms on a turf if the turf is considered to be plating or the atom is on level 2 (i.e. not hidden below floor tiles).
Also fixes a few types which didn't qdel() correctly when blown up.

Merge pull request #18218 from comma/slimefuckyou

Slimetoxin changes
Fixes midi2piano

Tick lag is now properly accounted for. This increases sheet music notes duration.

Resetting default params
Merge pull request #18282 from ARetardedWhale/patch-1

Industrial Veil loadout tweak
De-stealths hyposprays and autoinjectors


Merge pull request #18224 from Casper3667/hoodies

adds hoods to hoodies
Merge pull request #18264 from Hubblenaut/fixes

Fixes Material Duplication Glitch
Merge pull request #18287 from AtomicWorm/dev

De-stealths hyposprays and autoinjectors
Fixes maintenance drones

Merge pull request #18293 from comma/fixies

Small fixes
Adds limit to organ recovery.

Merge pull request #18286 from EditorRUS/dev

Fixes midi2piano
Gives APCs directional sprites


Merge pull request #18276 from PsiOmegaDelta/170816-ItIsSafeDownBelow

Adjusts when explosions affect below-floor objects.
Merge pull request #18295 from AtomicWorm/muh-perspective

Gives APCs directional sprites.
Fixes seed packets double-initing

Also moves code that depends on things existing into init.
Fixes rubber masks

Misspelled/mising icon states for skrell/human ones.
Shows mask's visible name again instead of always Unknown.
Updates changelog.

Merge pull request #18195 from FTangSteve/deathSpiral

Radiation damage and liver recovery surgery tweaks.
Merge pull request #18299 from PsiOmegaDelta/170819-Changelog

Updates changelog.
Underwear is now handled as items.

Your own underwear can be removed using the "Remove-Underwear" verb.
You can remove others underwear through the strip-menu.
To put on underwear, on yourself or someone else, click the target mob with the underwear in your active hand.
The wardrobe now offers a limited quota of underwear if replacements are needed.
You can increase this quota by returning underwear items.

Merge pull request #18292 from sabiram/itsadronesworld

Fixes a bunch of maintenance drone things
Merge pull request #18294 from PsiOmegaDelta/170818-PantySpace

Underwear is now handled as items.
IV Drips and Body Scanner Consoles are no longer dense


Critical diona nymph buffs.

Removed unnecessary vital flag from head organ.

Merge pull request #18263 from Ravensdale/iGotAnPaiInMyPocket

A couple pAI fixes
Revert "Merge pull request #18294 from PsiOmegaDelta/170818-PantySpace"

This reverts commit e91e339b33820869e3920c3697b67f75a11ea0df.

Revert "Revert "Merge pull request #18294 from PsiOmegaDelta/170818-PantySpace""

This reverts commit c2d5716522bdcca7628003561240f2dc077d6468.

Fixes popcorn recipe to need 5 salt

See title. Popcorn was destroying any enchilada ingredients because it took priority.
Adjusts visible organ messages on examine

Now it only displays message if SOMETHING is visible, no more 'something visible in the leg'.
Adds broken bones to the list for advanced diagnosis.
Non-damaged organs showing up too so you can tell if they're ok.
Adds both lobby screens

Adds both lobby screens (adds both lobby screens)
Vomit changes

I don't like how vomit currently is triggered because:
a) 5 damage is way to fucking low
b) it's weird that perforated lungs or stabbed hearts trigger vomiting.

So I came up with extremely complicated vomit scoring system to answer the question 'do we vomit now?'

First, it checks filtration organs, and adds their damage.
Missing organ is bad, but at some point (after broken damage) it becomes better to just not have it.

Having toxins in blood or radiation falls back to old total toxloss counting - we're assuming here that whatever damage there is is caused by those. Only using half of it, so unless filtration organs were hit it would take more than usual damage to start vomiting.

Alcohol and more alcohol also add bit to the score.

All in all you need at least 25 score to trigger vomiting.
This can be:
Bruised liver/kidneys (here and after where I say liver, kidneys work too).
Liver halfway to bruising, but you have alcohol in bloodstream.
A bit damaged liver, but you have A LOT of alcohol (toxic dose).
50+ overal toxloss that evaded liver and kidneys, but you have toxins in blood or radiation.

Reasoning is, aside from low threshold vomiting being extremely annoying, is that having it trigger on any organ's damage makes it useless for diagnostics. Other organs have distinct signs for their damage (low oxygen, blood not working etc), but liver and kidneys only had this blanket vomiting along with others. This change will make vomiting indicate most likely filtration systems failure, with overall toxins / rad poisoning being second possibility.
Makes non-dense object not count for explosion resistance

Mostly for open blastdoors to stop stopping tyhings.
Merge pull request #18273 from comma/hugeguts

Vomit changes
Changelog.

Merge pull request #18315 from nearlyNonexistent/popcorn-fix

Fixes Popcorn Recipe
Merge pull request #18316 from comma/thirdtimesthecharm

Adds both lobby screens
Merge pull request #18198 from Zero-Bits/reeda-booke

Fixes And Cleans Up Manuals
Merge pull request #18308 from Zuhayr/chirp

Critical diona nymph buffs.
Merge pull request #18296 from comma/pokeypokey

Adjusts visible organ messages on examine
Merge pull request #18304 from AtomicWorm/medical-QoL

A small change to IV drips and the Body Scanner console.
Merge pull request #18283 from Crushtoe/if-you-merge-this-you-are-uncool

Various Map Fixes
Merge pull request #18261 from comma/kablowie

Makes non-dense object not count for explosion resistance
Small fixes

Remove redundant proc and some deprecated code
Fixes MMI brain check


minifix

submited non updated file
Removes world.log use


run script


Fixes bluegoasts

Explosions make less of a mess with wires

Now only breaks wires if their turf is exposed.
Merge pull request #18253 from Hubblenaut/fixes

Fixes Bluegoasts
Merge pull request #18234 from Hubblenaut/pods

Fixes Second Deck Escape Pods not launching
Merge pull request #18251 from comma/a

Fixes MMI brain check
Merge pull request #18238 from comma/dokidokinicotine

Makes cigs less heart-racing
Fixes #8962 - messenger software gives PDA now

HOTFIX Follow Verb

Fixes globals?

Probably. This should be up to current standards
Fixes Material Duplication Glitch

Microwave sprite pattern fix.

Merge pull request #18265 from Earthcrusher/MicrowaveFix

Microwave sprite pattern fix.
Attempt at fixing multiz view

This uses a for loop to iterate through list, which I do not like that much, if there was a better way to truncate everything in a list to a plane it'd be neat.
Various Mapping Tweaks

Cleaned the Map

Merge pull request #18231 from Hubblenaut/hotfix2

HOTFIX Follow Verb
Merge pull request #18266 from Crushtoe/i-am-a-failure

Minor Mapping Tweaks and Teleporter Server Crashing Fix
Nerfs lung damage

At some point it was changed so ANY sort of damage makes lung give you 25% of max health in oxygen loss. This was confusing as  lung wasn't considered 'ruptured' by all the checks.
I bumped it back to ruptured lung, and instead made it give you a small amount of oxyloss just to indicate that something's off.
Sets limb blending to additive


Merge pull request #18274 from comma/coughcough

Nerfs lung damage
Keeps relative order

Should mostly fix everything, at this point I'd like reviewing code.
Merge pull request #18254 from comma/andstaydown

Explosions make less of a mess with wires
Industrial Veil loadout tweak

Just allows Taj to have industrial veils as Engineers, since it was missing before.
Fixes layering problems

Now everything is rendered properly, exception being edge cases of things that do not inform properly of changes (lowest turf being space or occasionally fire)
Various Map Fixes

Fixes Travis Issues

Merge pull request #18230 from tlc2013/dev

Updates the Lobby and Explosion art.
Fixes #10041 - pAI default language galcom

Fixes #10041 - Galcom default for all synths

Moves proc down to absolute pathing

Fixes indentation error

Adjusts when explosions affect sub-floor objects.

Explosions now only affect atoms on a turf if the turf is considered to be plating or the atom is on level 2 (i.e. not hidden below floor tiles).
Also fixes a few types which didn't qdel() correctly when blown up.

Merge pull request #18218 from comma/slimefuckyou

Slimetoxin changes
Fixes midi2piano

Tick lag is now properly accounted for. This increases sheet music notes duration.

Resetting default params
Merge pull request #18282 from ARetardedWhale/patch-1

Industrial Veil loadout tweak
Clears up plane code slightly, hotfixes lattice

Should work now, in theory supports up to 10 stacked z levels
De-stealths hyposprays and autoinjectors


Merge pull request #18224 from Casper3667/hoodies

adds hoods to hoodies
Merge pull request #18264 from Hubblenaut/fixes

Fixes Material Duplication Glitch
Merge pull request #18287 from AtomicWorm/dev

De-stealths hyposprays and autoinjectors
Fixes maintenance drones

Merge pull request #18293 from comma/fixies

Small fixes
Adds limit to organ recovery.

Merge pull request #18286 from EditorRUS/dev

Fixes midi2piano
Gives APCs directional sprites


Merge pull request #18276 from PsiOmegaDelta/170816-ItIsSafeDownBelow

Adjusts when explosions affect below-floor objects.
Merge pull request #18295 from AtomicWorm/muh-perspective

Gives APCs directional sprites.
Fixes seed packets double-initing

Also moves code that depends on things existing into init.
Fixes rubber masks

Misspelled/mising icon states for skrell/human ones.
Shows mask's visible name again instead of always Unknown.
Updates changelog.

Increased range

Because apparently there's more z levels than I thought
Changelog

Maybe this will let travis merge, I think I committed something odd before
Try fix CI

Perhaps
Basic conversion to subsystem

Code failures are probably still here if existed before.
Also unsure on proper wait.
Merge pull request #18195 from FTangSteve/deathSpiral

Radiation damage and liver recovery surgery tweaks.
Updates tests, removes some redundancies


Updates tests, removes some redundancies


Merge remote-tracking branch 'origin/OpenSpacePort' into OpenSpacePort


Merge pull request #18299 from PsiOmegaDelta/170819-Changelog

Updates changelog.
Underwear is now handled as items.

Your own underwear can be removed using the "Remove-Underwear" verb.
You can remove others underwear through the strip-menu.
To put on underwear, on yourself or someone else, click the target mob with the underwear in your active hand.
The wardrobe now offers a limited quota of underwear if replacements are needed.
You can increase this quota by returning underwear items.

Merge pull request #18292 from sabiram/itsadronesworld

Fixes a bunch of maintenance drone things
Removes excessive redundancy


Merge pull request #18294 from PsiOmegaDelta/170818-PantySpace

Underwear is now handled as items.
IV Drips and Body Scanner Consoles are no longer dense


Critical diona nymph buffs.

Removed unnecessary vital flag from head organ.

Merge pull request #18263 from Ravensdale/iGotAnPaiInMyPocket

A couple pAI fixes
Revert "Merge pull request #18294 from PsiOmegaDelta/170818-PantySpace"

This reverts commit e91e339b33820869e3920c3697b67f75a11ea0df.

Revert "Revert "Merge pull request #18294 from PsiOmegaDelta/170818-PantySpace""

This reverts commit c2d5716522bdcca7628003561240f2dc077d6468.

Fixes popcorn recipe to need 5 salt

See title. Popcorn was destroying any enchilada ingredients because it took priority.
Adjusts visible organ messages on examine

Now it only displays message if SOMETHING is visible, no more 'something visible in the leg'.
Adds broken bones to the list for advanced diagnosis.
Non-damaged organs showing up too so you can tell if they're ok.
Adds both lobby screens

Adds both lobby screens (adds both lobby screens)
Vomit changes

I don't like how vomit currently is triggered because:
a) 5 damage is way to fucking low
b) it's weird that perforated lungs or stabbed hearts trigger vomiting.

So I came up with extremely complicated vomit scoring system to answer the question 'do we vomit now?'

First, it checks filtration organs, and adds their damage.
Missing organ is bad, but at some point (after broken damage) it becomes better to just not have it.

Having toxins in blood or radiation falls back to old total toxloss counting - we're assuming here that whatever damage there is is caused by those. Only using half of it, so unless filtration organs were hit it would take more than usual damage to start vomiting.

Alcohol and more alcohol also add bit to the score.

All in all you need at least 25 score to trigger vomiting.
This can be:
Bruised liver/kidneys (here and after where I say liver, kidneys work too).
Liver halfway to bruising, but you have alcohol in bloodstream.
A bit damaged liver, but you have A LOT of alcohol (toxic dose).
50+ overal toxloss that evaded liver and kidneys, but you have toxins in blood or radiation.

Reasoning is, aside from low threshold vomiting being extremely annoying, is that having it trigger on any organ's damage makes it useless for diagnostics. Other organs have distinct signs for their damage (low oxygen, blood not working etc), but liver and kidneys only had this blanket vomiting along with others. This change will make vomiting indicate most likely filtration systems failure, with overall toxins / rad poisoning being second possibility.
Makes non-dense object not count for explosion resistance

Mostly for open blastdoors to stop stopping tyhings.
Merge pull request #18273 from comma/hugeguts

Vomit changes
Changelog.

Merge pull request #18315 from nearlyNonexistent/popcorn-fix

Fixes Popcorn Recipe
Merge pull request #18316 from comma/thirdtimesthecharm

Adds both lobby screens
Merge pull request #18198 from Zero-Bits/reeda-booke

Fixes And Cleans Up Manuals
Merge pull request #18308 from Zuhayr/chirp

Critical diona nymph buffs.
Merge pull request #18296 from comma/pokeypokey

Adjusts visible organ messages on examine
Merge pull request #18304 from AtomicWorm/medical-QoL

A small change to IV drips and the Body Scanner console.
Merge pull request #18283 from Crushtoe/if-you-merge-this-you-are-uncool

Various Map Fixes
Merge pull request #18261 from comma/kablowie

Makes non-dense object not count for explosion resistance
Merge branch for conflicts


Global var update

Just update test
Merge and cleanuo


Moves var

Seems the temporary list of turfs can be moved to fire() without consequences.
Fixes based on feedback

Removed name, unnecesary.
Update names properly

Now changing the character name will update the shadows too.
Converting to events [DNM]

This introduced a couple bugs, so may want to not merge this for now. Also tweaks stairs a bit.
Allows engineering module to respawn lights.

Corrects old reagent uses.

These were added during the refactoring and were missed.

Merge pull request #18386 from comma/saymyrank

Adds ability to set rank etc to syndie IDs
Fixes #18374

Init was called before the new was done.
Fixes miniscule amounts of alcohol givin you LIVER HURT OH GOD messages

It didn't do damage, but proc still sent a pain message, and there's no granurality between 0.1 and 30 damage there.
Added said granurality, 
made sure damage proc is only called when alcohol is in toxic dose, 
made low amount custom pain messages appear as warning span instead of bold red,
Made liver damage properly silent as comment suggests, it was inverted - 99% chance to be loud.
Tweaked fuel compressor to be borg compatible.

Merge pull request #18398 from Hubblenaut/dev

Locker Content Tweaks
Merge pull request #18408 from comma/ree

Fixes miniscule amounts of alcohol givin you LIVER HURT OH GOD messages
Merge pull request #18405 from comma/dev

Fixes #18374
Merge pull request #18409 from Zuhayr/borgrust

Tweaked fuel compressor to be borg compatible.
Bloody matters

Splits get_effective_blood_volume into three procs.
get_volume: straight up percentage of blood left
get_blood_circulation: what get_effective_blood_volume used to do, volume modified by heart condition
get_blood_oxygenation: thing that brain actually used, bu tthere was no way to see it anywhere in game.

Updates all usages to most relevant one.
Merge pull request #18395 from Casper3667/patch-2

Fixes pathfinder and explorer loadout.
Merge pull request #18400 from Zuhayr/dev

Allows engineering module to respawn lights.
Merge pull request #18376 from comma/superhotSUPERHOT

Per gas combustion product defines
Adjusts Pathfinder/Explorer access

They lose cargo-specific ones (bot and cargo access)
Instead they get xenoarch access to be able get to Excavation Prep. It doesn't let them get into NT areas since those use NT accesses too.
They also get access to Aquila so they can use any shuttle for the missions.
No need to double list all accesses now

Makes minimal_access default to access if not defined
Removes all the duplicated access lists from torch joblist.
Merge pull request #18399 from PsiOmegaDelta/170828-ReagentCorrections

Corrects old reagent uses.
Removes gotcha Faction var from jobs

It's literally not used anywhere except for making jobs not appear when you forget to set it.
Merge pull request #18414 from comma/backintown

Bloody matters
Merge pull request #18417 from comma/crushingfactions

Removes gotcha Faction var from jobs
Merge pull request #18415 from comma/assfinders

Everyone is opening Pathfinders PR why can't I open a Pathinfders PR
Bugfixes

Fixes some bugs introduced by latest changes
Merge


Update tests


Comment


Merge


Conflict fix


other file fix

BlueNexus reviewed (dismissed):
Going to need to see some profiler results before this'll get merged.

CrimsonShrike commented:
Profiling 

https://pastebin.com/Q70TnFvc

PsiOmegaDelta commented on code/controllers/subsystems/open_space_controller.dm:
```diff
@@ -0,0 +1,90 @@
+
+	for(last_object in turfs_to_process_old)
+		var/turf/T = last_object
+		if(!QDELETED(T))
```
A turf can never be deleted in BYOND, it's immediately replaced with another turf on the same ref.
Mostly just a minor optimization.

PsiOmegaDelta commented on code/controllers/subsystems/open_space_controller.dm:
```diff
@@ -0,0 +1,90 @@
+/datum/controller/subsystem/open_space/fire(resumed = 0)
+	// We use a different list so any additions to the update lists during a delay from CHECK_TICK
+	// don't cause things to be cut from the list without being updated.
+	var/list/turfs_to_process_old = turfs_to_process
```
You need to check if this is a resumed run or not. See https://github.com/tgstation/tgstation/blob/master/code/controllers/subsystem/processing/obj.dm#L12 for an example.
`turfs_to_process` corresponds to `processing` and `turfs_to_process_old` would correspond to `currentrun`.

If not, if the `turfs_to_process_old`-loop takes too long time any unprocessed turfs will never be processed.

Also note the use of 
```
if (MC_TICK_CHECK)
	return
```
when using roll-over lists like this.

PsiOmegaDelta commented on code/controllers/subsystems/open_space_controller.dm:
```diff
@@ -0,0 +1,90 @@
+
+/turf/simulated/open/Initialize()
+	. = ..()
+	if(GLOB.open_space_initialised)
```
Can instead return `INITIALIZE_HINT_LATELOAD` and implement `/turf/simulated/open/LateInitialize()`.
In this case you'll know that `/datum/controller/subsystem/open_space` is always ready and you'll probably not need `/datum/controller/subsystem/open_space/proc/initialize_open_space()` at all.

And if you still find some reason to keep that proc you should still do `TICK_CHECK` in the loop to avoid the lag.

PsiOmegaDelta commented on code/modules/multiz/turf.dm:
```diff
@@ -116,3 +166,34 @@
+
+/turf/simulated/open/proc/handle_move(var/atom/current_loc, var/atom/movable/am, var/atom/changed_loc)
+	//First handle objs and such
+	if(GLOB.open_space_initialised && !am.invisibility && isobj(am))
```
Your system doesn't seem to take possible invisibility changes into consideration.

comma replied:
Tracked now

PsiOmegaDelta reviewed (dismissed):
I've been pondering which objects should listen to which events but the current setup is probably, fine-ish.

PsiOmegaDelta commented on code/modules/multiz/zshadow.dm:
```diff
@@ -0,0 +1,103 @@
+			M.shadow.forceMove(OS)
+			M = M.shadow
+			OS = GetAbove(M)
+	// The topmost level does not need a shadow!
```
Then why does it even get one?

PsiOmegaDelta commented on code/controllers/subsystems/open_space_controller.dm:
```diff
@@ -0,0 +1,90 @@
+GLOBAL_DATUM_INIT(over_OS_darkness, /image, image('icons/turf/open_space.dmi', "black_open"))
+
+
+SUBSYSTEM_DEF(open_space)
```
And while you're (most likely) changing this to use `processing` and `currentrun` you should use `PROCESSING_SUBSYSTEM_DEF(open_space)` instead.
It'll pre-define these vars for you.

PsiOmegaDelta replied:
Actually, nevermind that change for now.
It relies on code I have not yet ported.

BlueNexus commented:
Merge commits.

CrimsonShrike commented:
Placing this temporarily on hold while I get new event merged in

CrimsonShrike committed:
Adds invisibility tracking to zshadows

Also cleans up one of the loops slightly

PromethTK11 commented:
rip

PurpleMartinJCK commented:
I guess let us know when this is ready for review again?

comma commented:
Pretty sure he mentioned it's ready on IRC

ghost commented:
Soon we'll finally have multi-z. 👌

(merge now pls)

CrimsonShrike commented:
Psi suggested some changes that I don't feel I'm able to do without rewriting a lot of code across a lot of files (namely watching all appareance vars like I did with invisibility). I can keep fixing merge conflicts, but I may have to give up if the feel is that the changes are needed (which they may, this copies Eris, but I am not sure how much of an issue that may or may not be there), since I don't see myself having time to do that in the near future.

CrimsonShrike committed:
Changed iteration order

Now will be iterated bottom-up, solving top turf being an update behind bottom one

CrimsonShrike commented:
Well at the very least now updating vertically is done correctly, boy, what a blunder that was.

comma commented:
@CrimsonShrike resolve the conflict, and I'll give it a green go.

re:last unhandled Psi's comment:
SUBSYSTEM_DEF(open_space) -> PROCESSING_SUBSYSTEM_DEF(open_space)
https://github.com/Baystation12/Baystation12/blob/dev/code/controllers/subsystems/processing/processing.dm
While this was up, processing SSes were ported, that do some things you defined yourself, would be nice to use them, but I'll understand that at this point you just want to get it in and maybe clean up at some point :V

comma reviewed (dismissed):
&& travis
